types/datum.go

```diff
@@ -1574,6 +1636,8 @@ func (d *Datum) convertToMysqlJSON(sc *stmtctx.StatementContext, target *FieldTy
 		}
 	case KindMysqlJSON:
 		ret = *d
+	case KindBinaryLiteral:
+		err = json.ErrInvalidJSONCharset.GenWithStackByArgs("binary")
```

Comment:
Please add a test for this case

Reply:
https://github.com/pingcap/tidb/pull/31031/files#diff-aa77b7b03663bb3d8eb4863157d94b2aa47f0ecec2e6876ea353eb825cadaf2fR219

Here it is :)